proof/invariant-abstract/KHeap_AI.thy

```diff
@@ -2620,4 +2659,200 @@ lemma update_sched_context_no_fail[wp]:
+  for no_cdt[wp]: "\<lambda>s. P (cdt s)"
+  and interrupt_states[wp]: "\<lambda>s. P (interrupt_states s)"
+
+lemma update_reply_mdb[wp]:
+  "update_reply ptr f \<lbrace>valid_mdb\<rbrace>"
+  by (wpsimp  wp: valid_mdb_lift)
+
+lemma update_reply_valid_idle[wp]:
+  "update_reply ptr f \<lbrace>valid_idle\<rbrace>"
+  apply (wpsimp simp: update_reply_def get_object_def)
+  apply (auto simp: valid_idle_def pred_tcb_at_def obj_at_def)
+  done
+
+crunch update_reply
+ for it[wp]: "\<lambda>s. P (idle_thread s)"
+
+lemma update_reply_only_idle[wp]:
+  "update_reply ptr f \<lbrace>only_idle\<rbrace>"
+  by (wpsimp wp: only_idle_lift)
+
+lemma update_reply_ifunsafe[wp]:
+  "update_reply ptr f \<lbrace>if_unsafe_then_cap\<rbrace>"
+  by (wpsimp simp: update_reply_def get_object_def obj_at_def is_reply)
+
+lemma update_reply_global_refs[wp]:
+  "update_reply ptr f \<lbrace>valid_global_refs\<rbrace>"
+  by (wpsimp simp: update_reply_def get_object_def wp: valid_global_refs_cte_lift)
+
+crunch update_reply
+ for irq_node[wp]: "\<lambda>s. P (interrupt_irq_node s)"
+
+lemma update_reply_valid_irq_node[wp]:
+  "update_reply ptr f \<lbrace>valid_irq_node\<rbrace>"
+  by (wpsimp wp: valid_irq_node_typ)
+
+lemma update_reply_valid_irq_handlers[wp]:
+  "update_reply ptr f \<lbrace>valid_irq_handlers\<rbrace>"
+  apply (wpsimp simp: update_reply_def set_object_def get_object_def valid_irq_handlers_def
+                      irq_issued_def ran_def)
+  apply (subgoal_tac "caps_of_state s (a, b) = Some cap")
+   apply fastforce
+  apply (subst cte_wp_caps_of_lift; auto simp: cte_wp_at_cases)
+  done
+
+lemma update_reply_kernel_window[wp]:
+  "update_reply ptr f \<lbrace>pspace_in_kernel_window\<rbrace>"
+  by (wpsimp simp: update_reply_def get_object_def
+               wp: set_object_pspace_in_kernel_window)
+
+lemma update_reply_cap_refs_kernel_window[wp]:
+  "update_reply ptr f \<lbrace>cap_refs_in_kernel_window\<rbrace>"
+  by (wpsimp simp: update_reply_def obj_at_def is_reply
+               wp: set_object_cap_refs_in_kernel_window get_object_wp)
+
+lemma update_reply_respect_device_region[wp]:
+  "update_reply ptr f \<lbrace>pspace_respects_device_region\<rbrace>"
+  by (wpsimp simp: update_reply_def get_object_def
+               wp: set_object_pspace_respects_device_region)
+
+lemma update_reply_cap_refs_respects_device_region[wp]:
+  "update_reply ptr f \<lbrace>cap_refs_respects_device_region\<rbrace>"
+  by (wpsimp simp: update_reply_def get_object_def obj_at_def is_reply
+               wp: set_object_cap_refs_respects_device_region)
+
+lemma update_reply_fault_tcbs_valid_states[wp]:
+  "update_reply ref f \<lbrace>fault_tcbs_valid_states\<rbrace>"
+  apply (wpsimp simp: update_reply_def get_object_def)
+  done
+
+lemma update_reply_cur_tcb[wp]:
+  "update_reply ptr f \<lbrace>cur_tcb\<rbrace>"
+  by (wpsimp simp: update_reply_def set_object_def get_object_def cur_tcb_def tcb_at_def get_tcb_def)
+
+lemma update_reply_cur_sc_tcb[wp]:
```

Comment:
I wonder if we have crunches and things for `update_sched_context` for some of these easy invariants, in which case maybe we can just add `update_reply` to the list of functions which we crunch. Again thinking about this in case I/we add an `update_endpoint` or `update_ntfn`